File: Polars.FSharp.Tests/UDFTests.fs
namespace Polars.FSharp.Tests

module UdfLogic =
    open Apache.Arrow

    // 场景 A: 类型转换 (Int32 -> String)
    let intToString (arr: IArrowArray) : IArrowArray =
        match arr with
        | :? Int64Array as i64Arr ->
            let builder = new StringViewArray.Builder()
            for i in 0 .. i64Arr.Length - 1 do
                if i64Arr.IsNull(i) then builder.AppendNull() |> ignore
                else 
                    let v = i64Arr.GetValue(i).Value
                    builder.Append $"Value: {v}" |> ignore
            builder.Build() :> IArrowArray

    // 为了兼容性保留 Int32
        | :? Int32Array as i32Arr ->
            let builder = new StringViewArray.Builder()
            for i in 0 .. i32Arr.Length - 1 do
                if i32Arr.IsNull i then 
                    builder.AppendNull() |> ignore
                else 
                    let v = i32Arr.GetValue(i).Value
                    builder.Append $"Value: {v}" |> ignore
            builder.Build() :> IArrowArray
        |_ -> failwith $"Expected Int32Array or Int64Array, but got: {arr.GetType().Name}"

    // 场景 B: 必定报错
    let alwaysFail (arr: IArrowArray) : IArrowArray =
        failwith "Boom! C# UDF Exploded!"

open Xunit
open Polars.FSharp
open Apache.Arrow
open System

type ``UDF Tests`` () =

    [<Fact>]
    member _.``Map UDF can change data type (Int -> String)`` () =
        // 1. 准备数据
        use csv = new TempCsv "num\n100\n200"
        let lf = LazyFrame.ScanCsv csv.Path
        
        // 2. 构造 C# 委托
        let udf = Func<IArrowArray, IArrowArray> UdfLogic.intToString

        // 3. 执行 Polars 查询
        // 关键点：必须传入 PlDataType.String，否则 Polars 可能会把结果当成 Int 处理导致乱码
        let df = 
            lf 
            |> pl.withColumnLazy (
                pl.col "num"
                |> fun e -> e.Map(udf, DataType.String)
                |> pl.alias "desc"
            )
            |> pl.selectLazy [ pl.col "desc" ]
            |> pl.collect

        // 4. 验证结果
        let arrowBatch = df.ToArrow()
        let strCol = arrowBatch.Column "desc" :?> StringViewArray
        
        Assert.Equal("Value: 100", strCol.GetString 0)
        Assert.Equal("Value: 200", strCol.GetString 1)

    [<Fact>]
    member _.``Map UDF error is propagated to F#`` () =
        // 1. 准备数据
        use csv = new TempCsv("num\n1")
        let lf = LazyFrame.ScanCsv csv.Path
        
        let udf = Func<IArrowArray, IArrowArray> UdfLogic.alwaysFail

        // 2. 断言会抛出异常
        let ex = Assert.Throws<Exception>(fun () -> 
            lf 
            |> pl.withColumnLazy (
                pl.col "num" 
                |> fun e -> e.Map(udf, DataType.SameAsInput)
            )
            // UDF 是 Lazy 执行的，只有 Collect/ToArrow 时才会触发
            |> pl.collect 
            |> ignore
        )

        // 3. 验证异常信息
        // 我们期望看到 C# 的异常信息包含在 PolarsError 里
        Assert.Contains("Boom! C# UDF Exploded!", ex.Message)
        Assert.Contains("C# UDF Failed", ex.Message) // 这是 Rust 代码里加的前缀

    [<Fact>]
    member _.``Generic Map UDF with Lambda (Int -> String)`` () =
        use csv = new TempCsv "num\n100\n"
        let lf = LazyFrame.ScanCsv csv.Path
        
        // --- 用户的代码极度简化 ---
        // 1. 定义一个简单的匿名函数 (int -> string)
        let myLogic = fun (x: int) -> sprintf "Num: %d" (x + 1)

        let df = 
            lf 
            |> pl.withColumnLazy (
                pl.col "num"
                // 2. 直接调用 Udf.map
                // 泛型 'T 和 'U 会自动推断为 int 和 string
                |> fun e -> e.Map(Udf.map myLogic, DataType.String) 
                |> pl.alias "res"
            )
            |> pl.selectLazy [ pl.col "res" ]
            |> pl.collect

        // 验证
        let arrow = df.ToArrow()
        let col = arrow.Column "res" :?> StringViewArray // 自动用了 StringView
        
        Assert.Equal("Num: 101", col.GetString 0)
        Assert.Equal(1, col.Length)
    [<Fact>]
    member _.``UDF: Map with Option (Null Handling)`` () =
        // 数据: [10, 20, null]
        // 注意: CSV 最后一行写 null 还是空行取决于解析器，这里用空行配合 Polars 默认行为
        use csv = new TempCsv "val\n10\n20\n" 
        let lf = LazyFrame.ScanCsv csv.Path

        // 逻辑: 
        // 输入是 int option
        // 如果有值且 > 15 -> 返回 Some (x * 2)
        // 否则 (<= 15 或原本就是 null) -> 返回 None (即 null)
        let logic (opt: int option) =
            match opt with
            | Some x when x > 15 -> Some (x * 2)
            | _ -> None

        let df = 
            lf 
            |> pl.withColumnLazy (
                pl.col "val"
                // 使用 mapOption，显式处理 Option 类型
                |> fun e -> e.Map(Udf.mapOption logic, DataType.Int32)
                |> pl.alias "res"
            )
            |> pl.collect

        // 验证
        let arrow = df.ToArrow()
        let col = arrow.Column "res" :?> Int32Array // 注意根据 DataType.Int32 生成的是 Int32Array

        // Row 0: 10 -> (<=15) -> None
        Assert.True(col.IsNull 0)

        // Row 1: 20 -> (>15) -> 40
        Assert.Equal(40, col.GetValue(1).Value)

        // Row 2: null -> (None) -> None
        Assert.True(col.IsNull 2)
    [<Fact>]
    member _.``UDF: Decimal Map (Series Native)`` () =
        // 1. 准备数据: String ["10.50", "20.25", null]
        let data = ["10.50"; "20.25"; null]
        let s = Series.create("str_vals", data)

        // 2. 先 Cast 成 Decimal(10, 2)
        // Series.Cast 直接返回一个新的 Series
        let sDec = s.Cast(DataType.Decimal(Some 10, Some 2))

        // 3. 定义业务逻辑
        // 输入: decimal option
        // 输出: decimal option (乘以 2)
        let logic (opt: decimal option) =
            opt |> Option.map (fun d -> d * 2m)

        // 4. 直接在 Series 上执行 Map!
        // 彻底告别 ToFrame -> withColumn -> Expr.Map 的绕路写法
        // 这里利用了我们刚加的 Series.Map 泛型重载 (内部自动调用 Udf.mapOption)
        let res = sDec.MapOption(logic, DataType.Decimal(Some 10, Some 2))

        // 5. 验证结果
        // 咱们现在的 Series.GetValue 已经非常智能了，直接取值即可
        
        // 10.50 * 2 = 21.00
        Assert.Equal(21.00m, res.GetValue<decimal> 0)
        
        // 20.25 * 2 = 40.50
        Assert.Equal(40.50m, res.GetValue<decimal> 1)
        
        // null -> null
        // 验证 Option 返回 None
        Assert.True(res.GetValue<decimal option>(2).IsNone)
    [<Fact>]
    member _.``Series: Map (Basic UDF)`` () =
        // Data: [1, 2, 3]
        let s = Series.create("nums", [1; 2; 3])

        // Logic: x * 10
        let doubleFunc (x: int) = x * 10
        
        // 1. 编译 UDF (类型推断: int -> int)
        let udf = Udf.map doubleFunc

        // 2. Apply to Series
        // 我们明确知道返回是 Int32
        let sRes = s.Map(udf, DataType.Int32)

        // 3. Verify
        Assert.Equal(10, sRes.GetValue<int> 0)
        Assert.Equal(30, sRes.GetValue<int> 2)

    [<Fact>]
    member _.``Series: Map (Option Handling)`` () =
        // Data: [10, null, 30]
        let s = Series.create("vals", [Some 10; None; Some 30])

        // Logic: Some(x) -> Some(x + 1), None -> Some(-1)
        // 这是一个只有 mapOption 才能做到的 "FillNull with Logic"
        let logic (opt: int option) =
            match opt with
            | Some x -> Some (x + 1)
            | None -> Some -1

        // Apply
        let sRes = s.Map(Udf.mapOption logic, DataType.Int32)

        // Verify
        Assert.Equal(11, sRes.GetValue<int> 0)
        Assert.Equal(-1, sRes.GetValue<int> 1) // None became -1

    [<Fact>]
    member _.``Series: Map (F# Lambda Sugar)`` () =
        // Data: ["a", "b"]
        let s = Series.create("txt", ["a"; "b"])

        // 直接传 lambda，享受 F# 的丝滑
        // 逻辑: string -> string (加后缀)
        // 注意：这里需要显式指明泛型参数，或者通过类型注解让编译器推断 'T 和 'U
        // 因为 Series 本身不知道自己存的是 string，所以 'T 必须匹配物理类型
        let sRes = s.Map<string, string>((fun x -> x + "_suffix"), DataType.String)
        
        Assert.Equal("a_suffix", sRes.GetValue<string> 0)

File: Polars.CSharp.Tests/UdfTests.cs
using Apache.Arrow;
using Polars.NET.Core;
using static Polars.CSharp.Polars; // 方便使用 Col, Lit

namespace Polars.CSharp.Tests;

// 1. 复刻 F# 的 UdfLogic 模块
public static class UdfLogic
{
    // 场景 A: 类型转换 (Int32/64 -> StringView)
    // 完全手动操作 Arrow，不依赖 UdfUtils
    public static IArrowArray IntToString(IArrowArray arr)
    {
        // F# match arr with | :? Int64Array ...
        if (arr is Int64Array i64Arr)
        {
            var builder = new StringViewArray.Builder();
            for (int i = 0; i < i64Arr.Length; i++)
            {
                if (i64Arr.IsNull(i))
                {
                    builder.AppendNull();
                }
                else
                {
                    long? v = i64Arr.GetValue(i);
                    builder.Append($"Value: {v}");
                }
            }
            return builder.Build();
        }
        
        if (arr is Int32Array i32Arr)
        {
            var builder = new StringViewArray.Builder();
            for (int i = 0; i < i32Arr.Length; i++)
            {
                if (i32Arr.IsNull(i))
                {
                    builder.AppendNull();
                }
                else
                {
                    int? v = i32Arr.GetValue(i);
                    builder.Append($"Value: {v}");
                }
            }
            return builder.Build();
        }

        throw new ArgumentException($"Expected Int32Array or Int64Array, but got: {arr.GetType().Name}");
    }

    // 场景 B: 必定报错
    public static IArrowArray AlwaysFail(IArrowArray arr)
    {
        throw new Exception("Boom! C# UDF Exploded!");
    }
    public static IArrowArray IntToDouble(IArrowArray arr)
    {
        if (arr is Int64Array i64Arr)
        {
            var builder = new DoubleArray.Builder();
            for (int i = 0; i < i64Arr.Length; i++)
            {
                if (i64Arr.IsNull(i)) 
                {
                    builder.AppendNull();
                }
                else 
                {
                    // 简单的数值计算
                    double val = i64Arr.GetValue(i).Value;
                    builder.Append(val / 2.0);
                }
            }
            return builder.Build();
        }
        // 为了兼容性处理 Int32
        if (arr is Int32Array i32Arr)
        {
            var builder = new DoubleArray.Builder();
            for (int i = 0; i < i32Arr.Length; i++)
            {
                if (i32Arr.IsNull(i)) builder.AppendNull();
                else builder.Append(i32Arr.GetValue(i).Value / 2.0);
            }
            return builder.Build();
        }

        throw new ArgumentException($"Expected Int32/64, got {arr.GetType().Name}");
    }
}

// 2. 复刻 F# 的测试类
public class UdfTests
{
    [Fact]
    public void Map_UDF_Memory_Data_Test()
    {
        // 1. 纯内存构造数据 (0..4, 共5行)
        // 使用 FromArrow 构造，绝对干净
        int rowCount = 5;
        var builder = new Int64Array.Builder();
        for (int i = 0; i < rowCount; i++) builder.Append(i * 10); // 0, 10, 20, 30, 40
        var arrowArray = builder.Build();

        using var df = DataFrame.FromArrow(
            new RecordBatch.Builder()
                .Append("num", false, col => col.Int64(arr => arr.AppendRange(Enumerable.Range(0, rowCount).Select(x => (long)x * 10))))
                .Build()
        );

        Assert.Equal(5, df.Height); // 确保源头是 5

        // 2. 转为 Lazy
        // using var lf = df.Lazy();

        // 3. 构造 UDF (Int64 -> Double)
        Func<IArrowArray, IArrowArray> udf = UdfLogic.IntToDouble;

        // 4. 执行
        using var res = df.Select(
            Col("num").Map(udf, DataType.Float64).Alias("res")
        );
        // 5. 验证
        // res.Show(); // 可以打开看看
        Assert.Equal(5, res.Height); // 应该是 5

        Assert.NotNull(res.Column("res"));
        Assert.Equal(5, res.Column("res").Length);
        Assert.Equal(0.0, res.Column("res").GetValue<double>(0)); // 0 / 2
        Assert.Equal(20.0, res.Column("res").GetValue<double>(4)); // 40 / 2
    }
    [Fact]
    public void Test_UDF_Map_Stable()
    {
        // 1. 准备数据 (5行)
        
        using var csv = new DisposableFile("num\n15\n25\n" +
                                          "35\n45\n55\n",".csv");
        using var df = DataFrame.ReadCsv(csv.Path);
        Assert.Equal(5, df.Height); // 确保数据是 5 行

        // 2. 构造 UDF (Int64 -> Int64)
        // 使用 Lambda，UdfUtils 会处理
        var udf = Col("num").Map<long, long>(x => x * 2, DataType.Int64).Alias("res");

        // 3. 执行
        using var res = df.Select(
            Col("num"),
            udf
        );
        // 看看是不是 3 行

        // 4. 验证
        Assert.Equal(5, res.Height); // 关键断言：行数不能丢
        
        Assert.Equal(30, res.Column("res").GetValue<long>(0));
        Assert.Equal(50, res.Column("res").GetValue<long>(1));
        Assert.Equal(70, res.Column("res").GetValue<long>(2));
    }
    [Fact]
    public void Map_UDF_Can_Change_Data_Type_Int_To_String()
    {
        // 1. 准备数据
        // F#: use csv = new TempCsv "num\n100\n200"
        using var csv = new DisposableFile("num\n100\n200\n",".csv");
        using var lf = LazyFrame.ScanCsv(csv.Path);

        // 2. 构造 C# 委托
        // F#: let udf = Func<IArrowArray, IArrowArray>(UdfLogic.intToString)
        Func<IArrowArray, IArrowArray> udf = UdfLogic.IntToString;

        // 3. 执行 Polars 查询
        using var df = lf.Select(
            Col("num")
            .Map(udf, DataType.String) 
            .Alias("desc")
        ).Collect();
        df.Show();
        // 4. 验证结果
        // F# let strCol = arrowBatch.Column "desc" :?> StringViewArray
        // 我们用 helper 或者直接 cast
        
        Assert.NotNull(df.Column("desc"));
        Assert.Equal("Value: 100", df.Column("desc").GetValue<string>(0)); // 使用之前修好的 GetStringValue
        Assert.Equal("Value: 200", df.Column("desc").GetValue<string>(1));
    }

    [Fact]
    public void Map_UDF_Error_Is_Propagated_To_CSharp()
    {
        // 1. 准备数据
        using var csv = new DisposableFile("num\n1",".csv");
        using var lf = LazyFrame.ScanCsv(csv.Path);

        Func<IArrowArray, IArrowArray> udf = UdfLogic.AlwaysFail;

        // 2. 断言会抛出异常
        var ex = Assert.Throws<PolarsException>(() => 
        {
            lf.Select(
                Col("num").Map(udf, DataType.SameAsInput)
            ).Collect();
        });

        // 3. 验证异常信息
        Assert.Contains("Boom! C# UDF Exploded!", ex.Message);
        // Assert.Contains("C# UDF Failed", ex.Message); // 取决于 Native 层是否加了前缀
    }
    [Fact]
    public void Test_UDF_HighLevel_Numeric()
    {
        // 场景 1: 简单的数值计算 (Int64 -> Int64)
        // 用户不需要知道 IArrowArray，只需要写 Lambda
        
        using var csv = new DisposableFile("num\n10\n20\n30\n",".csv");
        using var df = DataFrame.ReadCsv(csv.Path);

        // 定义逻辑: x * 2
        var doubleExpr = Col("num")
            .Map<long, long>(x => x * 2, DataType.Int64)
            .Alias("doubled");

        using var res = df.Select(Col("num"), doubleExpr);
        
        // 验证
        Assert.Equal(20, res.Column("doubled").GetValue<long>(0)); // 10 * 2
        Assert.Equal(60, res.Column("doubled").GetValue<long>(2)); // 30 * 2
    }

    [Fact]
    public void Test_UDF_HighLevel_String_Manipulation()
    {
        // 场景 2: 字符串处理 (String -> String)
        // 模拟业务逻辑：给名字加前缀
        
        using var csv = new DisposableFile("name\nAlice\nBob\n",".csv");
        using var df = DataFrame.ReadCsv(csv.Path);

        // 定义逻辑: "Hello, {name}!"
        var greetExpr = Col("name")
            .Map<string, string>(name => $"Hello, {name}!", DataType.String)
            .Alias("greeting");

        using var res = df.Select(Col("name"), greetExpr);
        
        // 验证
        Assert.Equal("Hello, Alice!", res.Column("greeting").GetValue<string>(0));
        Assert.Equal("Hello, Bob!", res.Column("greeting").GetValue<string>(1));
    }

    [Fact]
    public void Test_UDF_HighLevel_Type_Conversion()
    {
        // 场景 3: 跨类型转换 (Int64 -> String)
        // 这是最常用的场景之一，比如格式化数字
        
        using var csv = new DisposableFile("id\n1001\n1002\n",".csv");
        using var df = DataFrame.ReadCsv(csv.Path);

        // 定义逻辑: 将 ID 格式化为 "Order-#ID"
        // 输入是 long, 输出是 string
        var formatExpr = Col("id")
            .Map<long, string>(id => $"Order-{id}", DataType.String)
            .Alias("order_id");

        using var res = df.Select(Col("id"), formatExpr);
        
        // 验证     
        Assert.Equal("Order-1001", res.Column("order_id").GetValue<string>(0));
        Assert.Equal("Order-1002", res.Column("order_id").GetValue<string>(1));
    }
    [Fact]
    public void Test_UDF_Nullable_Output()
    {
        // 构造数据: 10, 0, 20
        // 我们想把 0 变成 Null
        using var csv = new DisposableFile("num\n10\n0\n20\n",".csv");
        using var df = DataFrame.ReadCsv(csv.Path);

        // 逻辑: 如果是 0 返回 null (C# null)，否则返回原值
        // 关键点：泛型参数是 <long, long?>
        var cleanExpr = Col("num")
            .Map<long, long?>(x => x == 0 ? null : x, DataType.Int64)
            .Alias("cleaned");

        using var res = df.Select(Col("num"), cleanExpr);
        
        // 验证
        Assert.Equal(10, res.Column("cleaned").GetValue<long>(0));
        Assert.Null(res.Column("cleaned").GetValue<long?>(1)); // 0 变成了 Null
        Assert.Equal(20, res.Column("cleaned").GetValue<long>(2));
    }
    [Fact]
    public void Test_UDF_Nullable_Input()
    {
        // 数据: 10, null
        using var csv = new DisposableFile("num\n10\n\n",".csv"); // 第二行是空
        using var df = DataFrame.ReadCsv(csv.Path);

        // 逻辑: 输入 int? -> 输出 string
        // 如果输入是 null，返回 "FoundNull"，否则返回 "Value:{x}"
        var checkNullExpr = Col("num")
            .Map<long?, string>(x => x.HasValue ? $"Value:{x}" : "FoundNull", DataType.String)
            .Alias("status");

        using var res = df.Select(Col("num"), checkNullExpr);
        
        Assert.Equal("Value:10", res.Column("status").GetValue<string>(0));
        Assert.Equal("FoundNull", res.Column("status").GetValue<string>(1)); // 成功捕获了 Null 输入！
    }
}
